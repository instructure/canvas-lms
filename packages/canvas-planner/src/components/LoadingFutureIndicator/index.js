/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that they will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React, { Component } from 'react';
import {bool, func, string} from 'prop-types';
import Button from '@instructure/ui-core/lib/components/Button';
import Container from '@instructure/ui-core/lib/components/Container';
import Spinner from '@instructure/ui-core/lib/components/Spinner';
import Text from '@instructure/ui-core/lib/components/Text';
import ErrorAlert from '../ErrorAlert';
import Waypoint from 'react-waypoint';
import formatMessage from '../../format-message';

export default class LoadingFutureIndicator extends Component {
  static propTypes = {
    loadingFuture: bool,
    allFutureItemsLoaded: bool,
    onLoadMore: func,
    loadingError: string,
    plannerActive: func,
  }

  static defaultProps = {
    loadingFuture: false,
    allFutureItemsLoaded: false,
    onLoadMore: () => {},
    loadingError: undefined,
    plannerActive: () => {return false;},
  }

  handleLoadMoreButton = () => {
    this.props.onLoadMore({});
  }

  handleWaypoint = () => {
    if (!this.props.loadingError && this.props.plannerActive()) {
      this.props.onLoadMore();
    }
  }

  renderLoadMore () {
    if (!this.props.loadingFuture && !this.props.allFutureItemsLoaded) {
      return <Button variant="link" onClick={this.handleLoadMoreButton}>
        {formatMessage('Load more')}
      </Button>;
    }
  }

  renderError () {
    if (this.props.loadingError) {
      // Show an Alert for the user, while including the underlying root cause error
      // in a hidden div in case we need to know what it was
      return (
        <div style={{width: '50%', margin: '0 auto'}}>
          <ErrorAlert error={this.props.loadingError}>
            {formatMessage('Error loading more items')}
          </ErrorAlert>
        </div>
      );
    }
  }

  renderLoading () {
    if (this.props.loadingFuture && !this.props.allFutureItemsLoaded) {
      return <Container>
        <Container display="inline">
          <Spinner size="small" margin="0 x-small 0 0" title={formatMessage('Loading...')} />
        </Container>
        <Text size="small" color="secondary">
          {formatMessage('Loading...')}</Text>
      </Container>;
    }
  }

  renderEverythingLoaded () {
    if (this.props.allFutureItemsLoaded) {
      return <Text color="secondary" size="small">
        {formatMessage('All items loaded')}</Text>;
    }
  }

  renderWaypoint () {
    if (!this.props.loadingFuture && !this.props.allFutureItemsLoaded){
      return <Waypoint onEnter={this.handleWaypoint} />;
    }
  }

  render () {
    return <div>
      {this.renderWaypoint()}
      <Container as="div" padding="x-large" textAlign="center">
        {this.renderError()}
        {this.renderLoadMore()}
        {this.renderLoading()}
        {this.renderEverythingLoaded()}
      </Container>
    </div>;
  }
}
