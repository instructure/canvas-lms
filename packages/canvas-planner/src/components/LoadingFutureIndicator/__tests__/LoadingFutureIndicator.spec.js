/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that they will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react';
import {shallow} from 'enzyme';
import LoadingFutureIndicator from '../index';

it('renders load more by default', () => {
  const wrapper = shallow(<LoadingFutureIndicator />);
  expect(wrapper).toMatchSnapshot();
});

it('renders loading when indicated', () => {
  const wrapper = shallow(<LoadingFutureIndicator loadingFuture />);
  expect(wrapper).toMatchSnapshot();
});

it('renders all future items loaded regardless of other props', () => {
  const wrapper = shallow(<LoadingFutureIndicator loadingFuture allFutureItemsLoaded />);
  expect(wrapper).toMatchSnapshot();
});

it('invokes the callback when the waypoint is triggered', () => {
  const mockLoad = jest.fn();
  const activeFunc = () => {return true;};
  const wrapper = shallow(<LoadingFutureIndicator onLoadMore={mockLoad} plannerActive={activeFunc} />);
  wrapper.instance().handleWaypoint();
  expect(mockLoad).toHaveBeenCalledWith();
});

it('does not invoke the callback when the waypoint is triggered, but the planner is not active', () => {
  const mockLoad = jest.fn();
  const activeFunc = () => {return false;};
  const wrapper = shallow(<LoadingFutureIndicator onLoadMore={mockLoad} plannerActive={activeFunc} />);
  wrapper.instance().handleWaypoint();
  expect(mockLoad.mock.calls.length).toBe(0);
});

it('invokes the callback when loading more button is clicked', () => {
  const mockLoad = jest.fn();
  const wrapper = shallow(<LoadingFutureIndicator onLoadMore={mockLoad} />);
  wrapper.find('Button').simulate('click');
  expect(mockLoad).toHaveBeenCalledWith({});
});

it('shows an Alert when there\'s a query error', () => {
  const wrapper = shallow(<LoadingFutureIndicator loadingError={"uh oh"}/>);
  expect(wrapper).toMatchSnapshot();
});
