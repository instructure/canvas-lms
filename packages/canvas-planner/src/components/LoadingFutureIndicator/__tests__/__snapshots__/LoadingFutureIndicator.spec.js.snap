// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders all future items loaded regardless of other props 1`] = `
<div>
  <Container
    as="div"
    display={null}
    padding="x-large"
    textAlign="center"
  >
    <Text
      as="span"
      color="secondary"
      letterSpacing="normal"
      size="small"
    >
      All items loaded
    </Text>
  </Container>
</div>
`;

exports[`renders load more by default 1`] = `
<div>
  <Waypoint
    bottomOffset="0px"
    fireOnRapidScroll={true}
    horizontal={false}
    onEnter={[Function]}
    onLeave={[Function]}
    onPositionChange={[Function]}
    topOffset="0px"
  />
  <Container
    as="div"
    display={null}
    padding="x-large"
    textAlign="center"
  >
    <Button
      as="button"
      buttonRef={[Function]}
      fluidWidth={false}
      margin="0"
      onClick={[Function]}
      size="medium"
      type="button"
      variant="link"
    >
      Load more
    </Button>
  </Container>
</div>
`;

exports[`renders loading when indicated 1`] = `
<div>
  <Container
    as="div"
    display={null}
    padding="x-large"
    textAlign="center"
  >
    <Container
      display={null}
    >
      <Container
        display="inline"
      >
        <Spinner
          margin="0 x-small 0 0"
          size="small"
          title="Loading..."
          variant="default"
        />
      </Container>
      <Text
        as="span"
        color="secondary"
        letterSpacing="normal"
        size="small"
      >
        Loading...
      </Text>
    </Container>
  </Container>
</div>
`;

exports[`shows an Alert when there's a query error 1`] = `
<div>
  <Waypoint
    bottomOffset="0px"
    fireOnRapidScroll={true}
    horizontal={false}
    onEnter={[Function]}
    onLeave={[Function]}
    onPositionChange={[Function]}
    topOffset="0px"
  />
  <Container
    as="div"
    display={null}
    padding="x-large"
    textAlign="center"
  >
    <div
      style={
        Object {
          "margin": "0 auto",
          "width": "50%",
        }
      }
    >
      <ErrorAlert
        error="uh oh"
      >
        Error loading more items
      </ErrorAlert>
    </div>
    <Button
      as="button"
      buttonRef={[Function]}
      fluidWidth={false}
      margin="0"
      onClick={[Function]}
      size="medium"
      type="button"
      variant="link"
    >
      Load more
    </Button>
  </Container>
</div>
`;
